### PR TITLE
Change `maybe_resize_storage_cpu` `new_size` arg to unsigned

### PR DESCRIPTION
**Pull request by malfet**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52672 `maybe_resize_storage_cuda` new_size argument should be unsigned
* **#52671 Change `maybe_resize_storage_cpu` `new_size` arg to unsigned**
* #52670 Fix wrong TORCH_CHECK usages

Code is written with the assumption that `new_size` is unsigned value,
and when function is called with negative value it silently returns a nullptr rather than raise an exception.
Fix abovementioned logic by converting new_size to unsigned type and let cpu_allocator raise exception on negative alloc.

Unroll nested if blocks by returning early if `new_size` is 0

Add `TestNN.test_adaptive_pooling_size_overflow` to indirecty validate the fix.

Fixes https://github.com/pytorch/pytorch/issues/50960

Differential Revision: [D26607549](https://our.internmc.facebook.com/intern/diff/D26607549)

### CODE REVIEW

**Review by walterddr — APPROVED** [review]

File: aten/src/ATen/native/Resize.h [hunk 1]
```diff
@@ -22,23 +22,24 @@ TORCH_API void resize_output(Tensor& output, IntArrayRef shape);
 // They are not in TH/THTensor.cpp because the at namespace is easier
 // to benchmark than TH; I can't get gbenchmark to call fns from THTensor.cpp
 
-static inline void maybe_resize_storage_cpu(TensorImpl* self, int64_t new_size) {
+static inline void maybe_resize_storage_cpu(TensorImpl* self, uint64_t new_size) {
```
Comment (mruberry) [inline aten/src/ATen/native/Resize.h:25]: It makes sense that this function would take an unsigned integer, but this update doesn't seem to reflect the name of the PR.

This function's also exclusively called by resize_impl_cpu_, isn't the issue that it computes `storage_size` in int64_t and doesn't check for the size being representable in int64_t? Naively, I guess I would have expected an explicit check for new_size being less than zero or zero. 

Also, in the case where new_size == 0, what happens if the tensor doesn't have a storage pointer?

Comment (malfet) [reply aten/src/ATen/native/Resize.h:25]: Will update PR title (and also skip the test under asan, which correctly detects integer overflow).

And there is already a check for `new_size == 0`, which exists, according to https://github.com/pytorch/pytorch/commit/c46c6a4fe68a8ed6a8adb35837e87209cf020547  handle zero sized tensor slicing, i.e.:
```
python -c "import numpy as np;import torch;t=torch.from_numpy(np.empty((0,4)));t[:,1::2]*=1"
```

Comment (mruberry) [reply aten/src/ATen/native/Resize.h:25]: I know there's a check for new_size == 0, but if the storage is uninitialized it just stays uninitialized in that case, right?

Comment (malfet) [reply aten/src/ATen/native/Resize.h:25]: Yes, as far as I understand, this is the intended behaviour.

Comment (mruberry) [reply aten/src/ATen/native/Resize.h:25]: OK.

**Pull request closed without merging**